test/extended/builds/build_pruning.go

```diff
@@ -60,234 +54,164 @@ var _ = g.Describe("[Feature:Builds][pruning] prune builds based on settings in
 					return false, err
 				}
-				if int32(len(builds.Items)) == *buildConfig.Spec.FailedBuildsHistoryLimit {
-					fmt.Fprintf(g.GinkgoWriter, "%v builds exist, retrying...", len(builds.Items))
-					return true, nil
+				for _, build := range builds.Items {
+					if build.Status.Phase == buildapi.BuildPhaseNew || build.Status.Phase == buildapi.BuildPhasePending || build.Status.Phase == buildapi.BuildPhaseRunning {
+						return false, nil
+					}
 				}
-				return false, nil
+				return true, nil
 			})
-
-			if err != nil {
-				fmt.Fprintf(g.GinkgoWriter, "%v", err)
-			}
-
-			passed := false
-			if int32(len(builds.Items)) == 2 || int32(len(builds.Items)) == 3 {
-				passed = true
-			}
-			o.Expect(passed).To(o.BeTrue(), "there should be 2-3 completed builds left after pruning, but instead there were %v", len(builds.Items))
-
-		})
-
-		g.It("should prune errored builds based on the failedBuildsHistoryLimit setting", func() {
-
-			g.By("creating test failed build config")
-			err := oc.Run("create").Args("-f", erroredBuildConfig).Execute()
 			o.Expect(err).NotTo(o.HaveOccurred())
 
-			g.By("starting four test builds")
-			for i := 0; i < 4; i++ {
-				br, _ := exutil.StartBuildAndWait(oc, "myphp")
-				br.AssertFailure()
-			}
-
-			buildConfig, err := oc.BuildClient().Build().BuildConfigs(oc.Namespace()).Get("myphp", metav1.GetOptions{})
-			if err != nil {
-				fmt.Fprintf(g.GinkgoWriter, "%v", err)
-			}
-
-			var builds *buildapi.BuildList
-
-			g.By("waiting up to one minute for pruning to complete")
+			g.By("waiting for pruning to complete")
 			err = wait.PollImmediate(pollingInterval, timeout, func() (bool, error) {
 				builds, err = oc.BuildClient().Build().Builds(oc.Namespace()).List(metav1.ListOptions{})
 				if err != nil {
 					fmt.Fprintf(g.GinkgoWriter, "%v", err)
 					return false, err
 				}
-				if int32(len(builds.Items)) == *buildConfig.Spec.FailedBuildsHistoryLimit {
-					fmt.Fprintf(g.GinkgoWriter, "%v builds exist, retrying...", len(builds.Items))
+				if int32(len(builds.Items)) <= (*bc.Spec.SuccessfulBuildsHistoryLimit + *bc.Spec.FailedBuildsHistoryLimit + 2) {
 					return true, nil
 				}
 				return false, nil
 			})
+			o.Expect(err).NotTo(o.HaveOccurred())
 
-			if err != nil {
-				fmt.Fprintf(g.GinkgoWriter, "%v", err)
-			}
+			g.By("checking that the correct number of builds exist")
+			builds, err = oc.BuildClient().Build().Builds(oc.Namespace()).List(metav1.ListOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
 
+			completedBuilds := 0
+			for _, build := range builds.Items {
+				e2e.Logf("Build Status: %s/%s", build.Name, build.Status.Phase)
+				if build.Status.Phase == buildapi.BuildPhaseComplete {
+					completedBuilds++
+				}
+			}
 			passed := false
-			if int32(len(builds.Items)) == 2 || int32(len(builds.Items)) == 3 {
+			if int32(completedBuilds) == *bc.Spec.SuccessfulBuildsHistoryLimit || int32(completedBuilds) == (*bc.Spec.SuccessfulBuildsHistoryLimit + 1) {
```

Comment:
remind me why are we expecting 2 or 3, not a specific number?  and put a comment in the test indicating why.

Reply:
Since the number of builds could be cached, our numbers could be off by 1, though this rewrite seems to be pretty consistently returning the correct number of builds.  Adding note in next push